internal/locate/region_cache.go

```diff
@@ -556,35 +680,46 @@ func NewRegionCache(pdClient pd.Client) *RegionCache {
 	c.tiflashComputeStoreMu.needReload = true
 	c.tiflashComputeStoreMu.stores = make([]*Store, 0)
 	c.notifyCheckCh = make(chan struct{}, 1)
-	c.ctx, c.cancelFunc = context.WithCancel(context.Background())
-	interval := config.GetGlobalConfig().StoresRefreshInterval
+	c.bg = newBackgroundRunner(context.Background())
+	c.enableForwarding = config.GetGlobalConfig().EnableForwarding
 
 	if config.GetGlobalConfig().EnablePreload {
 		logutil.BgLogger().Info("preload region index start")
-		if err := c.refreshRegionIndex(retry.NewBackofferWithVars(c.ctx, 20000, nil)); err != nil {
+		if err := c.refreshRegionIndex(retry.NewBackofferWithVars(c.bg.ctx, 20000, nil)); err != nil {
 			logutil.BgLogger().Error("refresh region index failed", zap.Error(err))
 		}
 		logutil.BgLogger().Info("preload region index finish")
 	} else {
 		c.mu = *newRegionIndexMu(nil)
 	}
 
-	// TODO(zyguan): refine management of background cron jobs
-	c.wg.Add(1)
-	go c.asyncCheckAndResolveLoop(time.Duration(interval) * time.Second)
-	c.enableForwarding = config.GetGlobalConfig().EnableForwarding
-	// Default use 15s as the update inerval.
-	c.wg.Add(1)
-	go c.asyncUpdateStoreSlowScore(time.Duration(interval/4) * time.Second)
-	if config.GetGlobalConfig().RegionsRefreshInterval > 0 {
-		c.timelyRefreshCache(config.GetGlobalConfig().RegionsRefreshInterval)
+	var (
+		refreshStoreInterval = config.GetGlobalConfig().StoresRefreshInterval
+		needCheckStores      []*Store
+	)
+	c.bg.scheduleWithTrigger(func(ctx context.Context, t time.Time) bool {
+		filter := func(state resolveState) bool {
+			return state != unresolved && state != tombstone && state != deleted
+		}
+		if t.IsZero() {
```

Comment:
We could comment this is triggered by store events.

Reply:
I've put a comment at https://github.com/tikv/client-go/pull/1171/files#diff-07678f6dc4c3b8185168ba1e2c221da085c85e26b518cf52a288e5af726cdfdcR593 , do we need to add another comment here?

Reply:
I think it's better to make it more clear.

Reply:
Done